File: backend/Services/FeedbackService.cs
using System.Text.Json;
using AiServiceApi.Data;
using AiServiceApi.Models;
using Microsoft.EntityFrameworkCore;

namespace AiServiceApi.Services;

/// <summary>
/// 反馈服务接口
/// </summary>
public interface IFeedbackService
{
    /// <summary>
    /// 创建反馈
    /// </summary>
    Task<CreateFeedbackResponse> CreateFeedbackAsync(string userId, CreateFeedbackRequest request);
    
    /// <summary>
    /// 获取用户的反馈列表
    /// </summary>
    Task<UserFeedbackListResponse> GetUserFeedbacksAsync(string userId, int page = 1, int pageSize = 20);
    
    /// <summary>
    /// 获取所有反馈列表（管理员）
    /// </summary>
    Task<UserFeedbackListResponse> GetAllFeedbacksAsync(
        FeedbackStatus? status = null, 
        FeedbackType? type = null,
        int page = 1, 
        int pageSize = 20);
    
    /// <summary>
    /// 获取反馈详情
    /// </summary>
    Task<FeedbackDetailInfo?> GetFeedbackDetailAsync(Guid feedbackId);
    
    /// <summary>
    /// 管理员回复反馈
    /// </summary>
    Task<RespondFeedbackResponse> RespondFeedbackAsync(string adminId, RespondFeedbackRequest request);
    
    /// <summary>
    /// 更新反馈状态
    /// </summary>
    Task<RespondFeedbackResponse> UpdateFeedbackStatusAsync(Guid feedbackId, FeedbackStatus status);
    
    /// <summary>
    /// 获取反馈统计信息
    /// </summary>
    Task<FeedbackStatisticsResponse> GetFeedbackStatisticsAsync();
}

/// <summary>
/// 反馈服务实现
/// </summary>
public class FeedbackService : IFeedbackService
{
    private readonly AppDbContext _dbContext;
    private readonly ILogger<FeedbackService> _logger;
    
    public FeedbackService(AppDbContext dbContext, ILogger<FeedbackService> logger)
    {
        _dbContext = dbContext;
        _logger = logger;
    }
    
    /// <summary>
    /// 创建反馈
    /// </summary>
    public async Task<CreateFeedbackResponse> CreateFeedbackAsync(string userId, CreateFeedbackRequest request)
    {
        try
        {
            var feedback = new UserFeedback
            {
                UserId = userId,
                Type = request.Type,
                Title = request.Title,
                Content = request.Content,
                RelatedModule = request.RelatedModule,
                Status = FeedbackStatus.Pending,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };
            
            // 处理截图（如果有）
            if (request.Screenshots?.Count > 0)
            {
                // 限制最多5张截图
                var screenshots = request.Screenshots.Take(5).ToList();
                feedback.Screenshots = JsonSerializer.Serialize(screenshots);
            }
            
            _dbContext.UserFeedbacks.Add(feedback);
            await _dbContext.SaveChangesAsync();
            
            _logger.LogInformation("User {UserId} created feedback {FeedbackId}: {Title}", 
                userId, feedback.Id, feedback.Title);
            
            return new CreateFeedbackResponse
            {
                Success = true,
                Message = "感谢您的反馈！我们会尽快处理。",
                FeedbackId = feedback.Id
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating feedback for user {UserId}", userId);
            return new CreateFeedbackResponse
            {
                Success = false,
                Error = "提交反馈失败，请稍后重试"
            };
        }
    }
    
    /// <summary>
    /// 获取用户的反馈列表
    /// </summary>
    public async Task<UserFeedbackListResponse> GetUserFeedbacksAsync(string userId, int page = 1, int pageSize = 20)
    {
        try
        {
            var query = _dbContext.UserFeedbacks
                .Include(f => f.User)
                .Where(f => f.UserId == userId)
                .OrderByDescending(f => f.CreatedAt);
            
            var totalCount = await query.CountAsync();
            
            var feedbacks = await query
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .Select(f => MapToDetailInfo(f))
                .ToListAsync();
            
            return new UserFeedbackListResponse
            {
                Success = true,
                Feedbacks = feedbacks,
                TotalCount = totalCount
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting feedbacks for user {UserId}", userId);
            return new UserFeedbackListResponse
            {
                Success = false,
                Error = "获取反馈列表失败"
            };
        }
    }
    
    /// <summary>
    /// 获取所有反馈列表（管理员）
    /// </summary>
    public async Task<UserFeedbackListResponse> GetAllFeedbacksAsync(
        FeedbackStatus? status = null, 
        FeedbackType? type = null,
        int page = 1, 
        int pageSize = 20)
    {
        try
        {
            var query = _dbContext.UserFeedbacks
                .Include(f => f.User)
                .AsQueryable();
            
            if (status.HasValue)
            {
                query = query.Where(f => f.Status == status.Value);
            }
            
            if (type.HasValue)
            {
                query = query.Where(f => f.Type == type.Value);
            }
            
            query = query.OrderByDescending(f => f.CreatedAt);
            
            var totalCount = await query.CountAsync();
            
            var feedbacks = await query
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .Select(f => MapToDetailInfo(f))
                .ToListAsync();
            
            return new UserFeedbackListResponse
            {
                Success = true,
                Feedbacks = feedbacks,
                TotalCount = totalCount
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting all feedbacks");
            return new UserFeedbackListResponse
            {
                Success = false,
                Error = "获取反馈列表失败"
            };
        }
    }
    
    /// <summary>
    /// 获取反馈详情
    /// </summary>
    public async Task<FeedbackDetailInfo?> GetFeedbackDetailAsync(Guid feedbackId)
    {
        var feedback = await _dbContext.UserFeedbacks
            .Include(f => f.User)
            .FirstOrDefaultAsync(f => f.Id == feedbackId);
        
        return feedback == null ? null : MapToDetailInfo(feedback);
    }
    
    /// <summary>
    /// 管理员回复反馈
    /// </summary>
    public async Task<RespondFeedbackResponse> RespondFeedbackAsync(string adminId, RespondFeedbackRequest request)
    {
        try
        {
            var feedback = await _dbContext.UserFeedbacks.FindAsync(request.FeedbackId);
            if (feedback == null)
            {
                return new RespondFeedbackResponse
                {
                    Success = false,
                    Error = "反馈不存在"
                };
            }
            
            feedback.AdminResponse = request.Response;
            feedback.RespondedAt = DateTime.UtcNow;
            feedback.RespondedBy = adminId;
            feedback.UpdatedAt = DateTime.UtcNow;
            
            if (request.NewStatus.HasValue)
            {
                feedback.Status = request.NewStatus.Value;
            }
            else
            {
                // 默认设置为处理中
                feedback.Status = FeedbackStatus.InProgress;
            }
            
            await _dbContext.SaveChangesAsync();
            
            _logger.LogInformation("Admin {AdminId} responded to feedback {FeedbackId}", adminId, request.FeedbackId);
            
            return new RespondFeedbackResponse
            {
                Success = true,
                Message = "回复成功"
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error responding to feedback {FeedbackId}", request.FeedbackId);
            return new RespondFeedbackResponse
            {
                Success = false,
                Error = "回复失败，请稍后重试"
            };
        }
    }
    
    /// <summary>
    /// 更新反馈状态
    /// </summary>
    public async Task<RespondFeedbackResponse> UpdateFeedbackStatusAsync(Guid feedbackId, FeedbackStatus status)
    {
        try
        {
            var feedback = await _dbContext.UserFeedbacks.FindAsync(feedbackId);
            if (feedback == null)
            {
                return new RespondFeedbackResponse
                {
                    Success = false,
                    Error = "反馈不存在"
                };
            }
            
            feedback.Status = status;
            feedback.UpdatedAt = DateTime.UtcNow;
            
            await _dbContext.SaveChangesAsync();
            
            return new RespondFeedbackResponse
            {
                Success = true,
                Message = "状态更新成功"
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating feedback status {FeedbackId}", feedbackId);
            return new RespondFeedbackResponse
            {
                Success = false,
                Error = "状态更新失败"
            };
        }
    }
    
    /// <summary>
    /// 获取反馈统计信息
    /// </summary>
    public async Task<FeedbackStatisticsResponse> GetFeedbackStatisticsAsync()
    {
        try
        {
            var totalCount = await _dbContext.UserFeedbacks.CountAsync();
            var pendingCount = await _dbContext.UserFeedbacks.CountAsync(f => f.Status == FeedbackStatus.Pending);
            var inProgressCount = await _dbContext.UserFeedbacks.CountAsync(f => f.Status == FeedbackStatus.InProgress);
            var resolvedCount = await _dbContext.UserFeedbacks.CountAsync(f => f.Status == FeedbackStatus.Resolved);
            var closedCount = await _dbContext.UserFeedbacks.CountAsync(f => f.Status == FeedbackStatus.Closed);
            
            var byType = await _dbContext.UserFeedbacks
                .GroupBy(f => f.Type)
                .Select(g => new { Type = g.Key, Count = g.Count() })
                .ToDictionaryAsync(x => x.Type.ToString(), x => x.Count);
            
            return new FeedbackStatisticsResponse
            {
                Success = true,
                Statistics = new FeedbackStatistics
                {
                    TotalCount = totalCount,
                    PendingCount = pendingCount,
                    InProgressCount = inProgressCount,
                    ResolvedCount = resolvedCount,
                    ClosedCount = closedCount,
                    ByType = byType
                }
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting feedback statistics");
            return new FeedbackStatisticsResponse
            {
                Success = false,
                Error = "获取统计信息失败"
            };
        }
    }
    
    #region Private Methods
    
    private static FeedbackDetailInfo MapToDetailInfo(UserFeedback feedback)
    {
        List<string>? screenshots = null;
        if (!string.IsNullOrEmpty(feedback.Screenshots))
        {
            try
            {
                screenshots = JsonSerializer.Deserialize<List<string>>(feedback.Screenshots);
            }
            catch { /* ignore parse errors */ }
        }
        
        return new FeedbackDetailInfo
        {
            Id = feedback.Id,
            UserId = feedback.UserId,
            UserEmail = feedback.User?.Email ?? "Unknown",
            UserNickname = feedback.User?.Nickname,
            Type = feedback.Type,
            Title = feedback.Title,
            Content = feedback.Content,
            RelatedModule = feedback.RelatedModule,
            Screenshots = screenshots,
            Status = feedback.Status,
            AdminResponse = feedback.AdminResponse,
            RespondedAt = feedback.RespondedAt,
            CreatedAt = feedback.CreatedAt,
            UpdatedAt = feedback.UpdatedAt
        };
    }
    
    #endregion
}

File: frontend/src/services/feedbackService.ts
import { getApiBaseUrl } from './config';
import { authStorage } from './authService';

// 获取 API 基础 URL
const getApiUrl = () => getApiBaseUrl();

/**
 * 反馈类型枚举
 */
export enum FeedbackType {
  Bug = 0,
  FeatureRequest = 1,
  Experience = 2,
  Other = 3,
}

/**
 * 反馈类型名称映射
 */
export const FeedbackTypeNames: Record<FeedbackType, string> = {
  [FeedbackType.Bug]: 'Bug报告',
  [FeedbackType.FeatureRequest]: '功能建议',
  [FeedbackType.Experience]: '使用体验',
  [FeedbackType.Other]: '其他',
};

/**
 * 反馈状态枚举
 */
export enum FeedbackStatus {
  Pending = 0,
  InProgress = 1,
  Resolved = 2,
  Closed = 3,
}

/**
 * 反馈状态名称映射
 */
export const FeedbackStatusNames: Record<FeedbackStatus, string> = {
  [FeedbackStatus.Pending]: '待处理',
  [FeedbackStatus.InProgress]: '处理中',
  [FeedbackStatus.Resolved]: '已解决',
  [FeedbackStatus.Closed]: '已关闭',
};

/**
 * 功能模块枚举（与后端对应）
 */
export enum FeatureModule {
  Chat = 1,
  Image = 2,
  Speech = 3,
  Legal = 4,
  Mystic = 5,
  RagChat = 6,
  Admin = 100,
  Other = 999,
}

/**
 * 功能模块名称映射
 */
export const FeatureModuleNames: Record<FeatureModule, string> = {
  [FeatureModule.Chat]: '聊天',
  [FeatureModule.Image]: '图片生成',
  [FeatureModule.Speech]: '语音',
  [FeatureModule.Legal]: '法律助手',
  [FeatureModule.Mystic]: '玄学助手',
  [FeatureModule.RagChat]: 'RAG聊天',
  [FeatureModule.Admin]: '管理功能',
  [FeatureModule.Other]: '其他',
};

/**
 * 创建反馈请求
 */
export interface CreateFeedbackRequest {
  type: FeedbackType;
  title: string;
  content: string;
  relatedModule?: FeatureModule;
  screenshots?: string[];
}

/**
 * 创建反馈响应
 */
export interface CreateFeedbackResponse {
  success: boolean;
  message?: string;
  error?: string;
  feedbackId?: string;
}

/**
 * 反馈详情信息
 */
export interface FeedbackDetail {
  id: string;
  userId: string;
  userEmail: string;
  userNickname?: string;
  type: FeedbackType;
  typeName: string;
  title: string;
  content: string;
  relatedModule?: FeatureModule;
  relatedModuleName?: string;
  screenshots?: string[];
  status: FeedbackStatus;
  statusName: string;
  adminResponse?: string;
  respondedAt?: string;
  createdAt: string;
  updatedAt: string;
}

/**
 * 反馈列表响应
 */
export interface FeedbackListResponse {
  success: boolean;
  error?: string;
  feedbacks: FeedbackDetail[];
  totalCount: number;
}

/**
 * 回复反馈请求
 */
export interface RespondFeedbackRequest {
  feedbackId: string;
  response: string;
  newStatus?: FeedbackStatus;
}

/**
 * 回复反馈响应
 */
export interface RespondFeedbackResponse {
  success: boolean;
  message?: string;
  error?: string;
}

/**
 * 反馈统计信息
 */
export interface FeedbackStatistics {
  totalCount: number;
  pendingCount: number;
  inProgressCount: number;
  resolvedCount: number;
  closedCount: number;
  byType: Record<string, number>;
}

/**
 * 反馈统计响应
 */
export interface FeedbackStatisticsResponse {
  success: boolean;
  error?: string;
  statistics?: FeedbackStatistics;
}

/**
 * 提交反馈
 */
export const createFeedback = async (
  request: CreateFeedbackRequest
): Promise<CreateFeedbackResponse> => {
  const token = authStorage.getToken();
  if (!token) {
    return { success: false, error: '请先登录' };
  }

  try {
    const response = await fetch(`${getApiUrl()}/api/feedback`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      return {
        success: false,
        error: errorData.error || '提交反馈失败',
      };
    }

    const data = await response.json();
    return data;
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : '提交反馈失败',
    };
  }
};

/**
 * 获取当前用户的反馈列表
 */
export const getMyFeedbacks = async (
  page: number = 1,
  pageSize: number = 20
): Promise<FeedbackListResponse> => {
  const token = authStorage.getToken();
  if (!token) {
    return { success: false, error: '请先登录', feedbacks: [], totalCount: 0 };
  }

  try {
    const response = await fetch(
      `${getApiUrl()}/api/feedback/my?page=${page}&pageSize=${pageSize}`,
      {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${token}`,
        },
      }
    );

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      return {
        success: false,
        error: errorData.error || '获取反馈列表失败',
        feedbacks: [],
        totalCount: 0,
      };
    }

    const data = await response.json();
    return data;
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : '获取反馈列表失败',
      feedbacks: [],
      totalCount: 0,
    };
  }
};

/**
 * 获取反馈详情
 */
export const getFeedbackDetail = async (
  feedbackId: string
): Promise<FeedbackDetail | null> => {
  const token = authStorage.getToken();
  if (!token) {
    return null;
  }

  try {
    const response = await fetch(`${getApiUrl()}/api/feedback/${feedbackId}`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      return null;
    }

    const data = await response.json();
    return data;
  } catch {
    return null;
  }
};

/**
 * 获取所有反馈列表（管理员）
 */
export const getAllFeedbacks = async (
  status?: FeedbackStatus,
  type?: FeedbackType,
  page: number = 1,
  pageSize: number = 20
): Promise<FeedbackListResponse> => {
  const token = authStorage.getToken();
  if (!token) {
    return { success: false, error: '请先登录', feedbacks: [], totalCount: 0 };
  }

  let url = `${getApiUrl()}/api/feedback/all?page=${page}&pageSize=${pageSize}`;
  if (status !== undefined) {
    url += `&status=${status}`;
  }
  if (type !== undefined) {
    url += `&type=${type}`;
  }

  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      return {
        success: false,
        error: errorData.error || '获取反馈列表失败',
        feedbacks: [],
        totalCount: 0,
      };
    }

    const data = await response.json();
    return data;
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : '获取反馈列表失败',
      feedbacks: [],
      totalCount: 0,
    };
  }
};

/**
 * 回复反馈（管理员）
 */
export const respondFeedback = async (
  request: RespondFeedbackRequest
): Promise<RespondFeedbackResponse> => {
  const token = authStorage.getToken();
  if (!token) {
    return { success: false, error: '请先登录' };
  }

  try {
    const response = await fetch(`${getApiUrl()}/api/feedback/respond`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      return {
        success: false,
        error: errorData.error || '回复反馈失败',
      };
    }

    const data = await response.json();
    return data;
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : '回复反馈失败',
    };
  }
};

/**
 * 更新反馈状态（管理员）
 */
export const updateFeedbackStatus = async (
  feedbackId: string,
  status: FeedbackStatus
): Promise<RespondFeedbackResponse> => {
  const token = authStorage.getToken();
  if (!token) {
    return { success: false, error: '请先登录' };
  }

  try {
    const response = await fetch(
      `${getApiUrl()}/api/feedback/${feedbackId}/status`,
      {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        body: JSON.stringify({ feedbackId, status }),
      }
    );

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      return {
        success: false,
        error: errorData.error || '更新状态失败',
      };
    }

    const data = await response.json();
    return data;
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : '更新状态失败',
    };
  }
};

/**
 * 获取反馈统计信息（管理员）
 */
export const getFeedbackStatistics = async (): Promise<FeedbackStatisticsResponse> => {
  const token = authStorage.getToken();
  if (!token) {
    return { success: false, error: '请先登录' };
  }

  try {
    const response = await fetch(`${getApiUrl()}/api/feedback/statistics`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      return {
        success: false,
        error: errorData.error || '获取统计信息失败',
      };
    }

    const data = await response.json();
    return data;
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : '获取统计信息失败',
    };
  }
};

/**
 * 格式化日期
 */
export const formatDate = (dateString: string): string => {
  if (!dateString) return '未知';
  
  const date = new Date(dateString);
  return date.toLocaleString('zh-CN', {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
  });
};
